provider/gateway/rest/router.go

```diff
@@ -340,6 +380,95 @@ func wsLogWriter(ctx context.Context, ws *websocket.Conn, cfg wsLogsConfig) {
 	}()
 }
 
+func wsEventWriter(ctx context.Context, ws *websocket.Conn, cfg wsEventsConfig) {
+	pingTicker := time.NewTicker(pingPeriod)
+	cctx, cancel := context.WithCancel(ctx)
+	defer func() {
+		pingTicker.Stop()
+		cancel()
+		_ = ws.Close()
+	}()
+
+	evts, err := cfg.client.LeaseEvents(cctx, cfg.lid, cfg.follow)
+	if err != nil {
+		cfg.log.Error("couldn't fetch events", "error", err.Error())
+		err = ws.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocketInternalServerErrorCode, ""))
+		if err != nil {
+			cfg.log.Error("couldn't push control message through websocket", "error", err.Error())
+		}
+		return
+	}
+
+	if evts == nil {
+		err = ws.WriteMessage(
+			websocket.CloseMessage,
+			websocket.FormatCloseMessage(websocketLeaseNotFound, ""))
+		if err != nil {
+			cfg.log.Error("couldn't push control message through websocket", "error", err.Error())
+		}
+		return
+	}
+
+	defer evts.Stop()
+
+	if err = ws.SetReadDeadline(time.Time{}); err != nil {
```

Comment:
This is unusual, shouldn't this be setting a time in the future?

Reply:
Set initial deadline to 0 till ping timer fires